gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -0,0 +1,314 @@
+      Duration idleTimeout) {
+
+    Preconditions.checkNotNull(executor, "executor can't be null");
+    Preconditions.checkNotNull(clock, "clock can't be null");
+    Preconditions.checkNotNull(checkInterval, "checkInterval can't be null");
+    Preconditions.checkNotNull(idleTimeout, "checkInterval can't be null");
+
+    Preconditions.checkArgument(
+        Duration.ZERO.compareTo(checkInterval) < 0, "checkInterval must be > 0");
+
+    Preconditions.checkArgument(
+        Duration.ZERO.compareTo(idleTimeout) <= 0, "idleTimeout must be >= 0");
+
+    this.executor = executor;
+    this.clock = clock;
+    this.checkInterval = checkInterval;
+    this.idleTimeout = idleTimeout;
+  }
+
+  /** Schedules the timeout check thread. */
+  public void start() {
+    executor.scheduleAtFixedRate(
+        new Runnable() {
+          @Override
+          public void run() {
+            checkAll();
+          }
+        },
+        checkInterval.toMillis(),
+        checkInterval.toMillis(),
+        TimeUnit.MILLISECONDS);
+  }
+
+  /** Wraps the target observer with timing constraints. */
+  public ResponseObserver<ResponseT> watch(
+      ResponseObserver<ResponseT> innerObserver, Duration waitTimeout) {
+    Preconditions.checkNotNull(innerObserver, "innerObserver can't be null");
+    Preconditions.checkArgument(Duration.ZERO.compareTo(waitTimeout) <= 0, "waitTimeout must >= 0");
+
+    WatchdogStream stream = new WatchdogStream(innerObserver, waitTimeout);
+    openStreams.put(stream, VALUE_MARKER);
+    return stream;
+  }
+
+  @VisibleForTesting
+  void checkAll() {
+    Iterator<Entry<WatchdogStream, Object>> it = openStreams.entrySet().iterator();
+
+    while (it.hasNext()) {
+      WatchdogStream stream = it.next().getKey();
+      if (stream.cancelIfStale()) {
+        it.remove();
+      }
+    }
+  }
+
+  enum State {
+    /** Stream has been started, but doesn't have any outstanding requests. */
+    IDLE,
+    /** Stream is awaiting a response from upstream. */
+    WAITING,
+    /**
+     * Stream received a response from upstream, and is awaiting outerResponseObserver processing.
+     */
+    DELIVERING
+  }
+
+  class WatchdogStream extends StateCheckingResponseObserver<ResponseT> {
+    private final Object lock = new Object();
+
+    private final Duration waitTimeout;
+    private boolean hasStarted;
+    private boolean autoAutoFlowControl = true;
+
+    private final ResponseObserver<ResponseT> outerResponseObserver;
+    private StreamController innerController;
+
+    @GuardedBy("lock")
+    private State state = State.IDLE;
+
+    @GuardedBy("lock")
+    private int pendingCount = 0;
+
+    @GuardedBy("lock")
+    private long lastActivityAt;
+
+    private volatile Throwable error;
+
+    WatchdogStream(ResponseObserver<ResponseT> responseObserver, Duration waitTimeout) {
+      this.waitTimeout = waitTimeout;
+      this.outerResponseObserver = responseObserver;
+      this.lastActivityAt = clock.millisTime();
```

Comment:
Minor: please move this initialization to L161 to make it consistent with the other `@GuardedBy` fields.

Reply:
done